### DISCUSSION TITLE
Change input requirement based on router path?

### DISCUSSION THREAD

**Discussion by Extarys** [trpc/trpc#4409] [Q&A] [answered]
Wasn't sure how to title this.

I have a main route like so:
`/trpc/issues.list`
`UserId` is optional

```
list: publicProcedure
		.input(
			z.object({
				limit: z.number().min(1).max(100).nullish(),
				cursor: z.string().nullish(),
				userId: z.string().nullish(),
			}).nullish(),
		)
		.query(async ({ input }) => {
```

And I'd like to reuse all the same routes, with the exception that the UserID must be provided, to help prevent the wrong information being requested in the frontend:

`/trpc/user.issues.list`
`UserId` is required

```
list: publicProcedure
		.input(
			z.object({
				limit: z.number().min(1).max(100).nullish(),
				cursor: z.string().nullish(),
				userId: z.string(),
			})
		)
		.query(async ({ input }) => {
```

Having to use "/trpc/issues.list" to fetch specific users issues for example, it might be easy to forget to add the UserId and then all the issues become kind of public. Since I have multiple routes similar to this one, it would be nice to be able, somehow, to pass options or inside the router, detect how it was called and change the input requirement based on that.

Right now, reusing routers looks like this:
```
export const userRouter = router({
	issues: issuesRouter,
        ...
```

**Not sure how I can modify my `issuesRouter` so it's more versatile. Any tips?**

**Comment by KATT** [comment, marked as answer]
You might want to do input merging: https://trpc.io/docs/server/validators#input-merging

**Reply by Extarys** [reply]
Oh thanks, I missed that!

**Reply by Extarys** [reply]
Hey @KATT , is there a way to use input merging on an entire router instead of moving everything in a "controller" and rearrange routes? I have time today to play with this, but I can't find a way to add a `baseProcedure` to a whole router, so I still need to duplicate all routes from `issues` to the `user` router. I'm looking for a way to use the existing `issues` router and add specific inputs _only_ when called from the `users` router. (Leaving the `issues` router/input as is when called directly) Or I might be missing something obvious here :joy:
